src/Database/LSMTree/Internal.hs

```diff
@@ -1157,6 +1175,146 @@ mergeRuns resolve hfs hbio caching alloc runPaths mergeLevel runs = do
+-- simpler, as it is read-only.
+data Cursor m h = Cursor {
+      -- | Mutual exclusion, only a single thread can read from a cursor at a
+      -- given time.
+      cursorState :: !(StrictMVar m (CursorState m h))
+    }
+
+data CursorState m h =
+    CursorOpen !(CursorEnv m h)
+  | CursorClosed  -- ^ Calls to a closed cursor raise an exception.
+  | CursorDrained  -- ^ All data has been read, future reads are empty.
+
+data CursorEnv m h = CursorEnv {
+    -- === Session-inherited
+
+    -- | The session that this cursor belongs to.
+    --
+    -- NOTE: Consider using the 'cursorSessionEnv' field and helper functions
+    -- like 'cursorHasFS' instead of acquiring the session lock.
+    cursorSession     :: !(Session m h)
+    -- | Use this instead of 'cursorSession' for easy access. An open cursor may
+    -- assume that its session is open. A session's global resources, and
+    -- therefore resources that are inherited by the cursor, will only be
+    -- released once the session is sure that no cursors are open anymore.
+  , cursorSessionEnv  :: !(SessionEnv m h)
+
+    -- === Cursor-specific
+
+    -- | Session-unique identifier for this cursor.
+  , cursorId          :: !Word64
+  , cursorWriteBuffer :: !WriteBuffer
+  , cursorReaders     :: !(Readers.Readers (PrimState m) (Handle h))
+    -- | The runs held open by the cursor. We must remove a reference when the
+    -- cursor gets closed.
+  , cursorRuns        :: !(V.Vector (Run (PrimState m) (Handle h)))
+  }
+
+{-# SPECIALISE withCursor :: TableHandle IO h -> (Cursor IO h -> IO a) -> IO a #-}
+-- | See 'Database.LSMTree.Normal.withCursor'.
+withCursor ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => TableHandle m h
+  -> (Cursor m h -> m a)
+  -> m a
+withCursor th = bracket (newCursor th) closeCursor
+
+{-# SPECIALISE newCursor :: TableHandle IO h -> IO (Cursor IO h) #-}
+-- | See 'Database.LSMTree.Normal.newCursor'.
+newCursor ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => TableHandle m h
+  -> m (Cursor m h)
+newCursor th = withOpenTable th $ \thEnv -> do
+    let cursorSession = tableSession thEnv
+    let cursorSessionEnv = tableSessionEnv thEnv
+    let hfs = tableHasFS thEnv
+    let hbio = tableHasBlockIO thEnv
+    -- We acquire a read-lock on the session open-state to prevent races, see
+    -- 'sessionOpenTables'.
+    withOpenSession cursorSession $ \_ -> do
+      -- only put the real state once creation completed successfully
+      cursorState <- newMVar CursorClosed
+      let !cursor = Cursor cursorState
+      -- Make sure that if now an exception is raised (and all resources get
+      -- freed by the registry) the cursor state remains 'CursorClosed',
+      -- so the session (which might track the cursor) doesn't later find it
+      -- open and double frees resources.
+      modifyWithTempRegistry (pure CursorClosed) (void <$> swapMVar cursorState) $ \reg _ -> do
```

Comment:
Does the problem become clear? AFAICT the implementation of  `duplicate` is unsafe in that regard.

I had another approach before, which used `tempFree` instead of `modifyWithTempRegistry`, see commit "other approach to resource management".